### PR TITLE
tests for _select_needles()

### PR DESCRIPTION
**Pull request by t0mpr1c3**
Pulled out needle selection code from `AYABControl._cnfLine()` into a new method `_select_needles()` and added test code.

### CODE REVIEW

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 1]
```diff
@@ -90,18 +91,18 @@ def get_progress(self):
...
-        elif self.__knitting_mode == KnittingMode.CLASSIC_RIBBER_1.value and self._numColors > 2:
-            self.__knit_func = self._doublebed_multicol
-        elif self.__knitting_mode == KnittingMode.MIDDLECOLORSTWICE_RIBBER.value and self._numColors >= 2:
-            self.__knit_func = self._middlecoltwice
-        elif self.__knitting_mode == KnittingMode.HEARTOFPLUTO_RIBBER.value and self._numColors >= 2:
-            self.__knit_func = self._heartofpluto
-        elif self.__knitting_mode == KnittingMode.CIRCULAR_RIBBER.value: # not restricted to 2 colors
-            self.__knit_func = self._circular_ribber
+        if self._knitting_mode == KnittingMode.SINGLEBED.value and self._numColors == 2:
+            self._knit_func = self.__singlebed_2col
```
Comment (chris007de) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:95]: self.__singlebed_2col is unknown, this has to be self._singlebed_2col

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 2]
```diff
@@ -164,57 +165,29 @@ def __get_carriage_info(self, msg):
...
-
-            for col in range(0, self.__image.imgWidth()):
-                pxl = (self.__image.imageExpanded())[indexToSend][col]
-                # take the image offset into account
-                if pxl == 1 and sendBlankLine is False:
-                    pxlNumber = col + self.__image.imgStartNeedle()
-                    # TODO implement for generic machine width
-                    if 0 <= pxlNumber and pxlNumber < MACHINE_WIDTH:
-                        bits[pxlNumber] = True
+            imgRow, color, indexToSend, sendBlankLine, lastLine = self.__knit_func(lineNumber, imgHeight)
```
Comment (chris007de) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:187]: self.__knit_func is unkown, this has to be self._knit_func

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 3]
```diff
@@ -164,57 +165,29 @@ def __get_carriage_info(self, msg):
...
+            
+            # create bitarray
+            bits = self._choose_needles(color, indexToSend, sendBlankLine)
```
Comment (chris007de) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:190]: self._choose_needles is unkown, this has to be self._select_needles

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 4]
```diff
@@ -463,7 +451,7 @@ def knit(self, pImage, pOptions):
         if self._current_state == KnittingState.SETUP:
             self.__formerRequest = 0
             self.__lineBlock = 0
-            self.__image = pImage
+            self._image = pImage
             self._startLine = pImage.startLine()
 
             self._numColors = pOptions["num_colors"]
```
Comment (chris007de) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:457]: self.__knitting_mode has to be self._knitting_mode

File: src/main/python/ayab/plugins/ayab_plugin/ayab_image.py [hunk 5]
```diff
@@ -19,17 +19,19 @@
 #    https://github.com/AllYarnsAreBeautiful/ayab-desktop
 
 from PIL import Image
+from bitarray import bitarray
 import numpy as np
 
+MACHINE_WIDTH = 200
```
Comment (chris007de) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_image.py:25]: This is declared in ayab.py, ayab_control.py and test_ayab_control.py already. Perhaps we could put this into a dedicated .py and include it.

File: src/main/python/ayab/tests/test_ayab_control.py [hunk 6]
```diff
@@ -31,157 +35,180 @@ def test__singlebed_2col(self):
...
         ayab_control._startLine = 1
-        assert ayab_control._circular_ribber(12,3,9) == (0,0, 0,False,False)
+        assert ayab_control._circular_ribber(12, 3) == (0, 0, 0, False, False)
+
+    def test__select_needles(self):
+        ayab_control = AYABControl()
+        ayab_control._numColors = 3
+        ayab_control._startLine = 0
+        ayab_control._knitting_mode = 0
+        ayab_control._image = ayabImage(Image.new('P', (40,1)), 2)
```
Comment (chris007de) [inline src/main/python/ayab/tests/test_ayab_control.py:198]: I would love to have short comments describing the test cases here, as I think it is not too obvious what is happening in the following lines.

File: src/main/python/ayab/plugins/ayab_plugin/ayab_image.py (hunk 5 above)
Comment (t0mpr1c3) [reply src/main/python/ayab/plugins/ayab_plugin/ayab_image.py:25]: Yes, there are a couple of other global constants I would like to introduce as well. I will do that in a separate commit.

**Comment by t0mpr1c3** [comment]
Thanks for the write privileges, Chris. I'll try to keep the git history as clean as possible.

**Pull request closed without merging**